Userland/Libraries/LibWeb/CSS/StyleValues/BasicShapeStyleValue.cpp

```diff
@@ -9,14 +9,175 @@
+}
+
+Gfx::Path Inset::to_path(CSSPixelRect reference_box, Layout::Node const& node) const
+{
+    // FIXME: A pair of insets in either dimension that add up to more than the used dimension
+    // (such as left and right insets of 75% apiece) use the CSS Backgrounds 3 § 4.5 Overlapping Curves rules
+    // to proportionally reduce the inset effect to 100%.
+
+    auto top = inset_box.top().to_px(node, reference_box.height()).to_float();
+    auto right = reference_box.width().to_float() - inset_box.right().to_px(node, reference_box.width()).to_float();
+    auto bottom = reference_box.height().to_float() - inset_box.bottom().to_px(node, reference_box.height()).to_float();
+    auto left = inset_box.left().to_px(node, reference_box.width()).to_float();
+
+    return path_from_resolved_rect(top, right, bottom, left);
+}
+
+String Inset::to_string() const
+{
+    return MUST(String::formatted("inset({} {} {} {})", inset_box.top(), inset_box.right(), inset_box.bottom(), inset_box.left()));
+}
+
+Gfx::Path Xywh::to_path(CSSPixelRect reference_box, Layout::Node const& node) const
+{
+    auto top = y.to_px(node, reference_box.height()).to_float();
+    auto bottom = top + max(0.0f, height.to_px(node, reference_box.height()).to_float());
+    auto left = x.to_px(node, reference_box.width()).to_float();
+    auto right = left + max(0.0f, width.to_px(node, reference_box.width()).to_float());
+
+    return path_from_resolved_rect(top, right, bottom, left);
+}
+
+String Xywh::to_string() const
+{
+    return MUST(String::formatted("xywh({} {} {} {})", x, y, width, height));
+}
+
+Gfx::Path Rect::to_path(CSSPixelRect reference_box, Layout::Node const& node) const
+{
+    // An auto value makes the edge of the box coincide with the corresponding edge of the reference box:
+    // it’s equivalent to 0% as the first (top) or fourth (left) value, and equivalent to 100% as the second (right) or third (bottom) value.
+
+    auto top = box.top().is_auto() ? 0 : box.top().to_px(node, reference_box.height()).to_float();
+    auto right = box.right().is_auto() ? reference_box.width().to_float() : box.right().to_px(node, reference_box.width()).to_float();
+    auto bottom = box.bottom().is_auto() ? reference_box.height().to_float() : box.bottom().to_px(node, reference_box.height()).to_float();
+    auto left = box.left().is_auto() ? 0 : box.left().to_px(node, reference_box.width()).to_float();
+
+    // The second (right) and third (bottom) values are floored by the fourth (left) and second (top) values, respectively.
+    return path_from_resolved_rect(top, max(right, left), max(bottom, top), left);
+}
+
+String Rect::to_string() const
+{
+    return MUST(String::formatted("rect({} {} {} {})", box.top(), box.right(), box.bottom(), box.left()));
+}
+
+static String radius_to_string(Circle::Radius radius)
+{
+    if (radius.has<LengthPercentage>()) {
+        return radius.get<LengthPercentage>().to_string();
+    }
+    switch (radius.get<Circle::FitSide>()) {
```

Comment:
Places where you repeatedly check/get a value from a `Variant` can be expressed using `Variant::visit()`, which is like its version of a `switch` statement. It makes sure that every case is covered. Here, that'd be something like:

```c++
radius.visit(
	[](LengthPercentage const& length_percentage) { return length_percentage.to_string(); },
	[](Circle::FitSide const& side) {
		switch (side) {
		case Circle::ClosestSide:
			return "closest-side"_string;
		case Circle::FarthestSide:
			return "farthest-side"_string;
		}
	}
);

Reply:
Oh this is a _lot_ cleaner, thank you for pointing it out!